### PR TITLE
htlcswitch: fix empty commit sig and no commit sig

### PR DESCRIPTION
**Pull request by joostjager**
Builds on top of #3564.

This PR fixes two known bugs in the commitment tx update sequence:
* Sending an commit sig message that didn't sign any new updates ("empty commit sig")
* Not sending a commit sig message when it is expected.

In addition to that, several simplifications are carried through where values were tracked in the link that can also be obtained from the channel.

### CODE REVIEW

**Comment by halseth** [comment]
For reference: https://github.com/lightningnetwork/lnd/issues/1927

**Comment by joostjager** [comment]
@halseth's repro of empty commitSig: #3012

File: htlcswitch/link.go [hunk 1]
```diff
@@ -1079,35 +1059,11 @@ out:
 
 			break out
 
-		case <-l.logCommitTick:
```
Comment (cfromknecht) [inline htlcswitch/link.go]: does this mean that in certain scenarios the commit dance will be faster?

File: htlcswitch/link.go [hunk 2]
```diff
@@ -349,14 +340,6 @@ type channelLink struct {
 	// sub-systems with the latest set of active HTLC's on our channel.
 	htlcUpdates chan *contractcourt.ContractUpdate
 
-	// logCommitTimer is a timer which is sent upon if we go an interval
-	// without receiving/sending a commitment update. It's role is to
-	// ensure both chains converge to identical state in a timely manner.
-	//
-	// TODO(roasbeef): timer should be >> then RTT
-	logCommitTimer *time.Timer
```
Comment (cfromknecht) [inline htlcswitch/link.go]: nice to see these go!

File: htlcswitch/link_isolated_test.go [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+	"crypto/sha256"
+	"testing"
+	"time"
+
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+type linkTestContext struct {
```
Comment (cfromknecht) [inline htlcswitch/link_isolated_test.go]: 😍

Comment (joostjager) [reply htlcswitch/link_isolated_test.go]: Yes, every pr should have a small clean up component...

File: build/prefix_log.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+// Debug formats message using the default formats for its operands and writes
+// to log with LevelDebug.
+func (p *PrefixLog) Debug(v ...interface{}) {
+	p.log.Debug(v...)
+}
+
+// Info formats message using the default formats for its operands and writes to
+// log with LevelInfo.
+func (p *PrefixLog) Info(v ...interface{}) {
+	p.log.Info(v...)
```
[the remark this answers is not in the document]
Comment (joostjager) [reply build/prefix_log.go]: no, forgotten. fixed.

File: lnwallet/channel.go [hunk 5]
```diff
@@ -1416,6 +1422,7 @@ func NewLightningChannel(signer input.Signer,
 		Capacity:          state.Capacity,
 		LocalFundingKey:   state.LocalChanCfg.MultiSigKey.PubKey,
 		RemoteFundingKey:  state.RemoteChanCfg.MultiSigKey.PubKey,
+		log:               build.NewPrefixLog(logPrefix, walletLog),
```
[the remark this answers is not in the document]
Comment (joostjager) [reply lnwallet/channel.go]: Hehe. Ok, I know that I sometimes comment on commit structure in PRs and suggest some of them to be merged. This is particular the case for structs that are introduced in a separate commit but it is hard to review the code without any further context. In this case the prefix logger is an object that doesn't need much context to review (although this is a sliding scale) and therefore I made it a separate commit.

[the remark this answers is not in the document]
Comment (joostjager) [reply lnwallet/channel.go]: Added commits to also make the link logging consistent. Open question is what we think is the best prefix to use for link and channel? uint64, shortchan, chanpoint?

File: lnwallet/channel.go [hunk 6]
```diff
@@ -4151,24 +4151,29 @@ func (lc *LightningChannel) ReceiveNewCommitment(commitSig lnwire.Sig,
...
+
+	localUpdatesPending := lc.localUpdateLog.logIndex !=
+		lastRemoteCommit.ourMessageIndex
 
-	localUpdatesSynced := (lastLocalCommit.ourMessageIndex ==
-		lastRemoteCommit.ourMessageIndex)
+	remoteUpdatesPending := lastLocalCommit.theirMessageIndex !=
+		lastRemoteCommit.theirMessageIndex
+
+	allSigned := !localUpdatesPending && !remoteUpdatesPending
```
[the remark this answers is not in the document]
Comment (joostjager) [reply lnwallet/channel.go]: But it isn't duplicated. Here we check if we owe a signature, while in `ReceiveNewCommitment`, we check whether the remote party owed anything.

File: htlcswitch/link.go (hunk 2 above)
Comment (joostjager) [reply htlcswitch/link.go]: Yes indeed!

File: htlcswitch/link.go (hunk 1 above)
Comment (joostjager) [reply htlcswitch/link.go]: Yes!

File: htlcswitch/link_isolated_test.go [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+	"crypto/sha256"
+	"testing"
+	"time"
+
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// sendHtlcBobToAlice sends an HTLC from Bob to Alice, that pays to a preimage
```
Comment (halseth) [inline htlcswitch/link_isolated_test.go]: why this move?

File: htlcswitch/link.go [hunk 8]
```diff
@@ -502,11 +501,11 @@ func (l *channelLink) Start() error {
 // NOTE: Part of the ChannelLink interface.
 func (l *channelLink) Stop() {
 	if !atomic.CompareAndSwapInt32(&l.shutdown, 0, 1) {
-		log.Warnf("channel link(%v): already stopped", l)
+		l.log.Warn("already stopped")
```
Comment (halseth) [inline htlcswitch/link.go]: Looks like the logging changes can perfectly fine go in its own PR separate from the far more critical commit sig changes.

File: htlcswitch/link_test.go [hunk 9]
```diff
@@ -4696,11 +4696,12 @@ func TestChannelLinkNoEmptySig(t *testing.T) {
 	ctx.receiveRevAndAckAliceToBob()
 	ctx.sendRevAndAckBobToAlice()
 
-	// The situation now is that Alice still doesn't have her two htlcs on
-	// the local commit tx. Bob needs to send a new signature and Alice can
-	// only wait for that. However, Alice's log commit timer fires and Alice
-	// sends a commitment tx containing no updates. THIS SHOULD NOT HAPPEN!
-	ctx.receiveCommitSigAliceToBob(2)
+	// Allow some time for the log commit ticker to trigger for Alice.
+	time.Sleep(time.Second)
```
Comment (halseth) [inline htlcswitch/link_test.go]: can also add this `time` as an argument to `assertNoMsgFromAlice`, then we can select on it instead of sleeping here.

File: htlcswitch/link_isolated_test.go (hunk 7 above)
Comment (joostjager) [reply htlcswitch/link_isolated_test.go]: It is a preparation from creating a link test context. I wanted to keep the move to a different file isolated, to make it easy to see the real changes in the next commit.

File: htlcswitch/link.go (hunk 8 above)
Comment (joostjager) [reply htlcswitch/link.go]: Separate prs created and linked.

File: htlcswitch/link_test.go (hunk 9 above)
Comment (joostjager) [reply htlcswitch/link_test.go]: fixed

File: lnwallet/channel.go [hunk 10]
```diff
@@ -4152,24 +4152,29 @@ func (lc *LightningChannel) ReceiveNewCommitment(commitSig lnwire.Sig,
...
+// yet.
+func (lc *LightningChannel) AllUpdatesSigned() bool {
 	lc.RLock()
 	defer lc.RUnlock()
 
-	lastLocalCommit := lc.localCommitChain.tip()
 	lastRemoteCommit := lc.remoteCommitChain.tip()
+	lastLocalCommit := lc.localCommitChain.tip()
+
+	localUpdatesPending := lc.localUpdateLog.logIndex !=
```
Comment (halseth) [inline lnwallet/channel.go]: Sprinkle some comments in here.

File: lnwallet/channel.go [hunk 11]
```diff
@@ -4152,24 +4152,29 @@ func (lc *LightningChannel) ReceiveNewCommitment(commitSig lnwire.Sig,
...
-// FullySynced returns a boolean value reflecting if both commitment chains
-// (remote+local) are fully in sync. Both commitment chains are fully in sync
-// if the tip of each chain includes the latest committed changes from both
-// sides.
-func (lc *LightningChannel) FullySynced() bool {
+// AllUpdatesSigned returns a boolean value reflecting whether we need to send
+// out a commitment signature because there are outstanding local updates and/or
+// updates in the local commit tx that aren't reflected in the remote commit tx
+// yet.
+func (lc *LightningChannel) AllUpdatesSigned() bool {
```
Comment (halseth) [inline lnwallet/channel.go]: Name `OweCommitment`?

File: htlcswitch/link.go [hunk 12]
```diff
@@ -1009,11 +1009,14 @@ out:
 			break out
 		}
 
-		// If the previous event resulted in a non-empty
-		// batch, reinstate the batch ticker so that it can be
-		// cleared.
+		// If the previous event resulted in a non-empty batch, resume
+		// the batch ticker so that it can be cleared. Otherwise pause
+		// the ticker to prevent waking up the htlcManager while the
+		// batch is empty.
```
Comment (halseth) [inline htlcswitch/link.go]: Should we be using a timer instead?

File: htlcswitch/link_test.go [hunk 13]
```diff
@@ -5890,6 +5890,104 @@ func TestChannelLinkHoldInvoiceRestart(t *testing.T) {
...
+
+	// Allow some time for the log commit ticker to trigger for Alice.
+	time.Sleep(time.Second)
+
+	// Now that Bob revoked, Alice should send the sig she owes.
+	//
+	// THIS SHOULD NOT HAPPEN.
+	ctx.assertNoMsgFromAlice()
+
+	aliceLink.Stop()
```
Comment (halseth) [inline htlcswitch/link_test.go]: nit: defer stop

File: lnwallet/channel.go [hunk 14]
```diff
@@ -3989,6 +3989,20 @@ func (lc *LightningChannel) ReceiveNewCommitment(commitSig lnwire.Sig,
 	localACKedIndex := lc.remoteCommitChain.tail().ourMessageIndex
 	localHtlcIndex := lc.remoteCommitChain.tail().ourHtlcIndex
 
+	// Check for empty commit sig.
```
[the remark this answers is not in the document]
Comment (joostjager) [reply lnwallet/channel.go]: added

File: lnwallet/channel.go (hunk 11 above)
Comment (joostjager) [reply lnwallet/channel.go]: renamed

File: htlcswitch/link.go (hunk 12 above)
Comment (joostjager) [reply htlcswitch/link.go]: It could be a timer, but also the ticker is currently used pretty heavily in the unit tests that cover this code. So we'd need something similar to `time.After` that also allows manual triggering. I don't really want to extend the scope of this pr to include that.

File: htlcswitch/link.go [hunk 15]
```diff
@@ -1853,11 +1814,14 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
...
-		// added by processRemoteAdds.
-		if l.channel.PendingLocalUpdateCount() > 0 {
+		// The revocation window opened up. If there are pending local
+		// updates, try to update the commit tx. Pending updates could
+		// already have been present because of a previously failed
+		// update to the commit tx or freshly added in by
+		// processRemoteAdds. Also in case there are no local updates,
+		// but there are still remote updates that are not in the remote
+		// commit tx yet, send out an update.
+		if l.channel.OweCommitment() {
```
Comment (halseth) [inline htlcswitch/link.go]: nit: should this change really be in a separate commit? Seems orthogonal to removing the ticker.

Comment (joostjager) [reply htlcswitch/link.go]: It is in the same commit because the log commit ticker was 'fixing' the problem caused by just checking `l.channel.PendingLocalUpdateCount() > 0`.

Comment (halseth) [reply htlcswitch/link.go]: Definitely not a blocker, was just curious.

File: htlcswitch/link.go [hunk 16]
```diff
@@ -1014,11 +991,14 @@ out:
...
 
-		// If the previous event resulted in a non-empty
-		// batch, reinstate the batch ticker so that it can be
-		// cleared.
-		if l.batchCounter > 0 {
+		// If the previous event resulted in a non-empty batch, resume
+		// the batch ticker so that it can be cleared. Otherwise pause
+		// the ticker to prevent waking up the htlcManager while the
+		// batch is empty.
+		if l.channel.PendingLocalUpdateCount() > 0 {
```
Comment (cfromknecht) [inline htlcswitch/link.go]: this is certainly correct, but does this need to be executed on every iteration? not a blocker

File: lnwallet/channel.go [hunk 17]
```diff
@@ -4156,24 +4156,36 @@ func (lc *LightningChannel) ReceiveNewCommitment(commitSig lnwire.Sig,
...
-		lastRemoteCommit.ourMessageIndex)
+	// There are remote updates pending if their remote commitment tx (our
+	// local commitment tx) contains updates that we don't have added to our
+	// remote commitment tx yet.
+	remoteUpdatesPending := lastLocalCommit.theirMessageIndex !=
+		lastRemoteCommit.theirMessageIndex
+
+	// If any of the conditions above is true, we owe a commitment
+	// signature.
+	oweCommitment := localUpdatesPending || remoteUpdatesPending
```
Comment (cfromknecht) [inline lnwallet/channel.go]: i see now, the local and remote PoV are reversed. is it possible to consolidate this with the empty commit sig check via a helper method, e.g. `oweCommitment(local bool) bool`?

**Comment by joostjager** [comment]
PR is approved, but need to do interop testing with CL en Eclair

File: htlcswitch/link.go (hunk 16 above)
Comment (joostjager) [reply htlcswitch/link.go]: Maybe not, but it I put it there to be sure that the flow can never get stuck if we forget to set the timer somewhere. It may be a little defensive, but I think it is justified given the importance of this logic. I don't think there is significant performance impact of executing the check with every iteration.

File: lnwallet/channel.go (hunk 17 above)
Comment (joostjager) [reply lnwallet/channel.go]: consolidate in accordance with your suggestion

**Comment by joostjager** [comment]
Same test for Eclair 0.3.2-5ad3944, no problems observed.

```
#!/bin/bash

while true
do
        LNCLI="$HOME/lightninglabs/lnd/lncli-debug --network=regtest"
        ECLAIRCLI="$HOME/Downloads/eclair-cli.sh -a localhost:8081 -p foobar"
        ID1=`od -vAn -N4 -tu4 < /dev/urandom`
        ID2=`od -vAn -N4 -tu4 < /dev/urandom`

        BOLT1=`$ECLAIRCLI createinvoice --amountMsat=10000 --description="test" | jq -r .serialized`
        BOLT2=`$LNCLI addinvoice 10 | jq -r .pay_req`

        $LNCLI --network=regtest payinvoice -f $BOLT1
        $ECLAIRCLI payinvoice --invoice=$BOLT2
done
```

**Comment by joostjager** [comment]
Finally interop with `lnd 0.8`. Works too.

File: htlcswitch/link.go (hunk 16 above)
Comment (cfromknecht) [reply htlcswitch/link.go]: sounds good, agree with those comments

File: lnwallet/channel.go (hunk 17 above)
Comment (cfromknecht) [reply lnwallet/channel.go]: looks great, thank you!

**Comment by joostjager** [comment]
Rebased

File: lnwallet/channel.go [hunk 18]
```diff
@@ -4140,6 +4150,65 @@ func (lc *LightningChannel) ReceiveNewCommitment(commitSig lnwire.Sig,
 	return nil
 }
 
+// OweCommitment returns a boolean value reflecting whether we need to send
+// out a commitment signature because there are outstanding local updates and/or
+// updates in the local commit tx that aren't reflected in the remote commit tx
+// yet.
+func (lc *LightningChannel) OweCommitment(local bool) bool {
+	lc.RLock()
```
Comment (Roasbeef) [inline lnwallet/channel.go]: Perhaps we should instead obtain the mutex once within `ReceiveNewCommitment`, then either make this a private method, or update the `godoc` to note that it expects the mutex to already be held?

File: htlcswitch/link.go [hunk 19]
```diff
@@ -1109,19 +1112,10 @@ out:
 			}
 
 		case <-l.cfg.BatchTicker.Ticks():
-			// If the current batch is empty, then we have no work
```
Comment (Roasbeef) [inline htlcswitch/link.go]: Nice, this should also lower idle CPU usage for routing nodes with tons of channels.

File: lnwallet/channel.go (hunk 18 above)
Comment (joostjager) [reply lnwallet/channel.go]: It is also called without a lock from `ChannelLink`. I now created a private no-lock version and a public method that wraps it in a lock.

I am wondering whether our channel manipulations are always safe though. Something like this for example:

```
if lc.OweCommitment(true) {
			commitSig, htlcSigs, _, err := lc.SignNextCommitment()
```

I think the channel isn't modified concurrently, but why it the lock there then? Is it a safety precaution? Or for reading channel state in another subsystem?

**Comment by joostjager** [comment]
Test from #3012 added to this PR. It needed a bit of modification, but the idea of the scenario is the same.

For review, this is the delta with the previous revision of this pr:
https://github.com/joostjager/lnd/compare/hodl-drop-fix-better-ref..joostjager:hodl-drop-fix-better

File: lnwallet/channel.go (hunk 18 above)
Comment (Roasbeef) [reply lnwallet/channel.go]: The lock was mostly a safety precaution when the struct was first created just to ensure things were goroutine safe, so all other code added afterwards followed that trend. In the past there were also instances of grabbing live state machines from the peer, and using that to populate RPC information.
